Bugfix: Add debug option to log response headers

asein-sinch opened the pull request:
**How to enable debug logs:**
```
import winston from 'winston';
const winstonLogger = winston.createLogger({
  level: 'debug',
  transports: [new winston.transports.Console()],
  format: winston.format.logstash(),
});

new SinchClient({ projectId, keyId, keySecret, logger:  winstonLogger});
```

**Debug logs:**
```
[Sinch SDK][Debug][AvailableRegionsApi][ListAvailableRegions][401]
HTTP method: GET
URL: https://numbers.api.sinch.com/v1/projects/37b62a7b-0177-429a-bb0b-e10f848de0b8/availableRegions?types=LOCAL&types=MOBILE
Response Headers:  Headers {
  [Symbol(map)]: [Object: null prototype] {
    'www-authenticate': [
      'Bearer error="invalid_token", error_description="Jwt expired at 2023-10-02T14:57:08Z", error_uri="https://tools.ietf.org/html/rfc6750#section-3.1"'
    ],
    'cache-control': [ 'no-cache, no-store, max-age=0, must-revalidate' ],
    pragma: [ 'no-cache' ],
    expires: [ '0' ],
    'x-content-type-options': [ 'nosniff' ],
    'strict-transport-security': [ 'max-age=31536000 ; includeSubDomains' ],
    'x-frame-options': [ 'DENY' ],
    'x-xss-protection': [ '0' ],
    'referrer-policy': [ 'no-referrer' ],
    'x-envoy-upstream-service-time': [ '2' ],
    date: [ 'Tue, 02 Dec 2025 22:25:38 GMT' ],
    server: [ 'istio-envoy' ],
    connection: [ 'close' ],
    'content-length': [ '0' ]
  }
}

```

asein-sinch committed:
Bugfix: Add debug option to log response headers

asein-sinch committed:
Fix comment

asein-sinch committed:
Fix comments

asein-sinch commented on packages/sdk-client/src/plugins/exception/exception.response.ts:
```diff
@@ -75,11 +80,25 @@ export class ExceptionResponse<
-        context.response?.headers,
+    if (!context.response?.ok) {
+      this.logger.debug(
+        `[Sinch SDK][Debug][${context.apiName}][${context.operationId}][${context.response?.status}]\nHTTP method: ${context.requestOptions.method}\nURL: ${context.url}\nResponse Headers: ${this.formatHeaders(context.response?.headers)}`,
```
Done

JPPortier commented on packages/sdk-client/src/logger/logger.ts:
❤️ 
May be adding some unit tests to ensure proper/format and log level.
Not sure we want to go to this direction for console logger and this PR,  but other logging frameworks (e.g.: winston pointed by PR comment) are enabling to configure log level to not flood logs output/destination

Hurus111 committed:
Fix post-rebase logger integration issues in sdk-client.

Use global console for SinchLogger fallbacks so tests can spy on warnings, and drop unused OAuth2 token request logger after v1.5-next error handling.

JPPortier commented on packages/sms/src/rest/v1/sms-domain-api.ts:
```diff
@@ -59,7 +60,9 @@ export class SmsDomainApi implements Api {
       this.lazyClient.getApiClient();
     } catch (error) {
-      console.error('Impossible to assign the new credentials to the SMS API');
+      new SinchLogger(this.lazyClient.sharedConfig.logger ?? console).error(
```
Here we are assuming we have to send a log onto console even if configuration could have been, by purpose of SDK user, to NOT have any logger (multliple times)

What do you think :
- either we assume there is, by contract, a logger defined
- test if null/undefined and do not log anything

Hurus111 replied:
Fixed 🟡 
Removed all ?? console fallbacks; call sites use resolveLogger(...). logger: null → silent NOOP_LOGGER.
[packages/sms/src/rest/v1/sms-domain-api.ts](https://github.com/sinch/sinch-sdk-node/pull/219/files) (and all other *-service.ts / *-domain-api.ts)

JPPortier commented on packages/sdk-core/src/sinch-client.ts:
```diff
@@ -26,6 +26,9 @@ export class SinchClient {
    * @param {SinchClientParameters} params - The object containing the Sinch credentials.
    */
   constructor(params: SinchClientParameters) {
+    if (!params.logger) {
```
See comment about using or not a default logger.
What if for any purpose an end user do not want any log ? (thinking about a logging call causing an issue)
It won't be possible.
We could have distinction between `null` VS `undefined` to provide 
- null -> a dummy logger not logging anything
- undefined -> a default logger based onto init of SinchLogger

And, by contract we could remove all validation against not defined logger. e.g.: https://github.com/sinch/sinch-sdk-node/pull/219/changes#diff-85a0299065d0f2bd71e5151e1983edb3447d2d9770692d9f8ac007467246bc20R69

Hurus111 replied:
Fixed 🟡 
- null -> silent
- undefined -> default console

[packages/sdk-core/src/sinch-client.ts](https://github.com/sinch/sinch-sdk-node/pull/219/files) resolved once via resolveLogger() in SinchClient and each Service constructor.

JPPortier commented on packages/sdk-client/src/plugins/exception/exception.response.ts:
```diff
@@ -9,18 +10,27 @@ export class ExceptionResponse<
   ): PluginRunner<V | Record<string, unknown>, V> {
     return {
       transform: (res: V) => {
+        this.debug(context);
```
🤔
Not a big fan to have a dependency onto logger/SinchLogger here.
Finally: it is "just" a container.

We could think about who should be responsible of logging. 
Here, not sure a "simple" container should log itself something.
It is an exception response. Why not having transport layer performing the log itself and removing dependency onto any logger here?

Hurus111 replied:
Fixed 🟡 
Logger removed from ExceptionResponse. Failed HTTP debug logging moved to ApiFetchClient.
[packages/sdk-client/src/client/api-fetch-client.ts](https://github.com/sinch/sinch-sdk-node/pull/219/files)

JPPortier commented on packages/sdk-client/src/logger/logger.ts:
```diff
@@ -0,0 +1,27 @@
+  info(message: string, ...meta: any[]): void;
+  warn(message: string, ...meta: any[]): void;
+  error(message: string, ...meta: any[]): void;
+}
```
We could imagine, for performance/memory consumption purpose having interface defining a callback called by logger when the message will be really send.
Then; it will avoid to format message before calling logger which will finally decide to not log anything because of required severity is not reached.

Hurus111 replied:
Done 🟢 
Logger methods now accept either a plain `string` or a `callback () => string`. The SDK builds the message only when the logger actually emits it.

JPPortier commented on packages/sdk-client/src/domain/domain-interface.ts:
```diff
@@ -186,3 +188,8 @@ export type MailgunRegion = SupportedMailgunRegion | string;
   ...SupportedMailgunRegion,
 };
+
+export interface LoggerParameters {
```
Strictly speaking; this is nor defining "logger parameters" but the logger itself

Hurus111 replied:
Fixed 🟡 
Renamed to `WithLogger`

Hurus111 committed:
feat(sdk-client): add configurable logging primitives

Hurus111 committed:
refactor(sdk-client): move failed HTTP debug logging to ApiFetchClient

Hurus111 committed:
feat: resolve logger at entry points and remove console fallbacks

Hurus111 committed:
test(sdk-client): add logger tests and update service logging assertions

Hurus111 committed:
docs(sdk-client): changelog for v1.5.0 configurable logging

Hurus111 committed:
feat(sdk-client): add logger example script and update package.json

Hurus111 committed:
refactor(sdk-client): replace SinchLogger with direct logger calls in API classes

Hurus111 committed:
refactor(sdk-client): enhance resolveLogger to wrap loggers with SinchLogger

Hurus111 committed:
refactor(fax): simplify convertToSupportedFileType by removing logger and warning messages

Hurus111 committed:
refactor(sdk-client): remove SinchLogger usage in favor of direct logger calls across multiple files

JPPortier commented on examples/snippets/sdk-client/logger.js:
```diff
@@ -0,0 +1,43 @@
+ * The SDK accepts any object implementing the Logger interface (debug, info, warn, error).
+ * Pass it via the `logger` option on SinchClient.
+ *
+ * Example with Winston (install separately: npm install winston):
```
Adding a link onto Winston npmjs package or github repo ?

Hurus111 replied:
Added ` * Example with Winston (https://www.npmjs.com/package/winston — install separately: npm install winston):
`

JPPortier commented on examples/snippets/sdk-client/logger.js:
```diff
@@ -0,0 +1,43 @@
+ *     transports: [new winston.transports.Console()],
+ *     format: winston.format.logstash(),
+ *   });
+ *
```
Why not having this comment part of the snippet source itself ?
The snippet is to highlight how to add a custom snippet but finally we just init a client without any below

Hurus111 replied:
I wanted to avoid adding more dependencies to the project (even though it's only used here) @JPPortier

JPPortier commented on examples/snippets/sdk-client/logger.js:
```diff
@@ -0,0 +1,43 @@
+
+  // Omit `logger` to use console, pass `logger: null` to silence SDK output,
+  // or pass a custom logger — see the Winston example in the header comment.
+  const sinch = new SinchClient({ projectId, keyId, keySecret });
```
Provide here (see comment above) an init with a defined logger

Hurus111 replied:
Agreed the snippet should pass a logger in code: we'll use a minimal inline implementation so it runs without extra dependecies. If it's fine for you, Winston can stay in the comment as an optional integration example

Hurus111 replied:
Also added intentionally deprecated call to show custom log:
> [Sinch SDK][Warn] ** DEPRECATION NOTICE ** The "conversationRegion" property will become mandatory in the next major version of the SDK and not default to "us" anymore. Please set it to a valid region.
✅ SDK request completed.

JPPortier commented on packages/conversation/src/rest/v1/conversation-domain-api.ts:
```diff
@@ -67,7 +65,7 @@ export class ConversationDomainApi implements Api {
       this.lazyClient.getApiClient();
     } catch (error) {
-      new SinchLogger(resolveLogger(this.lazyClient.sharedConfig.logger)).error(
+      this.lazyClient.sharedConfig.logger!.error(
```
We should never receive a null/undefined `this.lazyClient.sharedConfig.logger` here.
If sinch client init is checking against:
- null: then a `NOOP_LOGGER` should have been created and in use 
- undefined: then a `CONSOLE_LOGGER` should have been created and in use

Hurus111 replied:
After init, `sharedConfig.logger` is always a resolved Logger via `ResolvedSinchClientParameters`. So no more `logger!` everywhere

JPPortier commented on packages/numbers/src/rest/v1/numbers-service.ts:
```diff
@@ -76,6 +77,7 @@ export class NumbersService {
    * @param {SinchClientParameters} params - an Object containing the necessary properties to initialize the service
    */
   constructor(params: SinchClientParameters) {
+    params.logger = resolveLogger(params.logger);
```
Here the input `params` parameter is modified itself.
Better to store result of resolvConf wihtin `sharedClient` right ?
(could be the cause of these errors: https://github.com/sinch/sinch-sdk-node/actions/runs/27703589896/job/81946190935)

Hurus111 replied:
The service no longer mutates caller params; it uses `resolveClientParameters` and stores the result on the `lazyClient.sharedConfig`

Hurus111 replied:
I also added a shared `LazyApiClient` base class in `@sinch/sdk-client` so each domain lazy client only implements `getApiClient()`

Hurus111 committed:
refactor(sdk-client): update logger example to use custom logger and improve documentation

Hurus111 committed:
feat(sdk-client): add resolveClientParameters function to ensure logger is always resolved

Hurus111 committed:
feat(sdk-client): add LazyApiClient class and export from index

Hurus111 committed:
refactor(sdk-client): replace optional logger chaining with direct logger calls across multiple API classes

Hurus111 committed:
refactor(sdk-client): change apiFetchClient property from protected to public in LazyApiClient class

Hurus111 committed:
refactor(tests): update hostname parameter assignment to use lazyClient.sharedConfig

JPPortier commented on packages/sdk-client/src/client/lazy-api-client.ts:
```diff
@@ -0,0 +1,17 @@
+  public apiFetchClient?: ApiFetchClient;
+  public sharedConfig: ResolvedSinchClientParameters;
+
+  constructor(params: SinchClientParameters | ResolvedSinchClientParameters) {
```
Is the `SinchClientParameters |` is still required ?
Regarding the service files, they perform a call to `const sharedConfig = resolveClientParameters(params);` before creating the underlying client, so the LazyClient will always receive a `ResolvedSinchClientParameters` right ?
e.g. from : https://github.com/sinch/sinch-sdk-node/pull/219/changes#diff-e08475f2da0aa9048bf9aaf2fa12ce8a08638fb91dbdec846c1276355d3905dbR127

JPPortier commented on packages/conversation/src/rest/v1/conversation-service.ts:
```diff
@@ -132,10 +124,11 @@ export class ConversationService {
    */
   constructor(params: SinchClientParameters) {
-    const sharedConversationClient = new LazyConversationApiClient(params);
+    const sharedConfig = resolveClientParameters(params);
```
In `setApiClientConfig` function just below, the variable is named `resolvedParams`
Better to have a common pattern/naming

JPPortier commented on packages/elastic-sip-trunking/src/rest/v1/elastic-sip-trunking-service.ts:
ConversationService (all all other services) are having a public `setApiClientConfig` function https://github.com/sinch/sinch-sdk-node/blob/4cf6337c9748afc78a4bfd3e6073516c917fa4ab/packages/conversation/src/rest/v1/conversation-service.ts#L148C10-L148C28 and a setCredentials.
Here only setCredentials. 
Why this special use case for EST ?

Hurus111 committed:
feat(sdk-client): integrate Winston logger into SDK client example and update package.json

Hurus111 committed:
refactor(sdk-client): restructure logger implementation by introducing logger-types and separating SinchLogger functionality

Hurus111 committed:
chore(sdk-client): update CHANGELOG for version 1.5.0 to clarify SDK logging feature enhancements

Hurus111 committed:
refactor(sdk-client): streamline logger resolution in API client options to enhance consistency and clarity

Hurus111 committed:
refactor(sdk-client): update API client initialization to use resolved parameters for consistency across services

Hurus111 committed:
refactor(tests): update API client initialization in test files to use resolveClientParameters for consistency
